ForceJump mode for NoFall

SmugNugg opened the pull request:
I implemented a ForceJump mode for NoFall.
This jumps just before the player will land.

SmugNugg committed:
ForceJump mode for NoFall

I implemented a ForceJump mode for NoFall.
This jumps just before the player will land.

SmugNugg committed:
Merge branch 'nextgen' of https://github.com/SmugNugg/LiquidBounce into nextgen

SmugNugg committed:
Improved code readability

Improved code readability, extended into multiple helper functions.

SmugNugg committed:
Added playeronly check in nopush.

superblaubeere27 commented on src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/entity/MixinLivingEntity.java:
```diff
@@ -139,13 +144,17 @@ private Vec3d hookFixRotation(Vec3d instance, double x, double y, double z) {
 
+
     @Inject(method = "pushAwayFrom", at = @At("HEAD"), cancellable = true)
-    private void hookNoPush(CallbackInfo callbackInfo) {
+    private void hookNoPush(Entity entity, CallbackInfo ci) {
         if (ModuleNoPush.INSTANCE.getEnabled()) {
-            callbackInfo.cancel();
+            // Checking if the entity is a playerEntity or a livingEntity depending if the playerOnly is toggled or not.
+            if ((entity instanceof PlayerEntity && ModuleNoPush.INSTANCE.getPlayerOnly()) ||
+                    (entity instanceof LivingEntity && !ModuleNoPush.INSTANCE.getPlayerOnly())) {
+                ci.cancel();
+            }
```
This should be it's own PR

SmugNugg replied:
Yes i already made a new pull request for it. 
The reason why this happened is because i didn't know how to make a new pull request sadly 😭 
But I've figured it out.

> This should be it's own PR

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/movement/ModuleNoPush.kt:
```diff
@@ -18,12 +18,21 @@
+import net.ccbluex.liquidbounce.utils.client.regular
+import net.ccbluex.liquidbounce.utils.entity.strafe
+import net.minecraft.network.packet.s2c.play.PlayerPositionLookS2CPacket
 
 /**
  * NoPush module
  *
  * Disables pushing from other players.
  */
-object ModuleNoPush : Module("NoPush", Category.MOVEMENT)
+object ModuleNoPush : Module("NoPush", Category.MOVEMENT) {
+    public var playerOnly by boolean("Player Only", true)
+}
```
This too

SmugNugg replied:
> This too

I know i was lost on how to make new pull requests

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/nofall/modes/NoFallForceJump.kt:
```diff
@@ -0,0 +1,79 @@
+ * This mode forces the player to jump just when his about to land,
+ * preventing fall damage.
+ */
+internal object NoFallForceJump : Choice("Force Jump") {
```
Spaces are added automatically by the frontend `ForceJump` -> `Force Jump`

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/nofall/modes/NoFallForceJump.kt:
```diff
@@ -0,0 +1,79 @@
+    override val parent: ChoiceConfigurable<*>
+        get() = ModuleNoFall.modes
+
+    private val blockDistance by int("Block Distance", 1, 0..5)
+    private val fallDistance by float("Fall Distance", 3.35f, 3.35f..10.0f)
+    private val jumpHeight by float("Jump Height", 0.42f, 0.1f..0.42f)
```
s.a.

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/nofall/modes/NoFallForceJump.kt:
```diff
@@ -0,0 +1,79 @@
+    val packetHandler = handler<PacketEvent> {
+        val packet = it.packet
+        if (packet is PlayerMoveC2SPacket && player.fallDistance > fallDistance) {
+            val blockBelowPos = calculateBlockBelowPosition()
+            val blockBelow: BlockState? = getBlockStateAtPosition(blockBelowPos)
+
+            if (!jumpTriggered && blockBelow != null && !blockBelow.isAir) {
```
It would be better (and simpler) if you did this check with `FallingPlayer` (see https://github.com/CCBlueX/LiquidBounce/blob/nextgen/src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/nofall/modes/NoFallMLG.kt#L178)

superblaubeere27 reviewed (changes requested):
Good idea, good work. I have some suggestions though

SmugNugg committed:
Update NoFallForceJump.kt

Made the force jump -> forcejump.
I am not sure if the usage with the FallingPlayer is correct.
So if incorrect please tell me

SmugNugg committed:
Update NoFallForceJump.kt

removed unused functions

SmugNugg reviewed (commented):
made it better by making request changes

1zun4 committed:
Revert "Added playeronly check in nopush."

This reverts commit 43eb8b5151d6564991017d6af35879c753aa4fc8.

1zun4 committed:
Merge branch 'refs/heads/nextgen' into fork/nextgen

1zun4 committed:
make it compile

1zun4 committed:
properly implemented

1zun4 commented:
@SmugNugg I made some changes to your PR because the original code not only did not compile, it did not work properly. I have now fixed this. Thanks for your PR and please try to use this as a learning approach :)

1zun4 reviewed (approved):
yeah i approve my own changes
